### ISSUE TITLE
New nullability warnings regarding lambda conversions are added in perfectly valid code upon updating to the latest compiler

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#57440]
**Version Used**: 
```
Branch main (26 Oct 2021)
Latest commit a4ec465 by Gen Lu:
Merge pull request #57338 from genlu/FixWarning

Fix 'AssemblyVersion overridden by auto-generated version' warning in build
```

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

delegate void D1<T>([NotNull] ref T? s);
[return: NotNull] delegate T D2<T>();

class C<T> where T : new()
{
  void Method()
  {
    D1<T> d1 = (ref T? s) => s = new T();
    D2<T> d2 = () => new T();
  }
}
```

**Expected Behavior**:
No warnings. 
The code doesn't have any warnings in C# 9 in the latest release version of VS.
The code doesn't violate any nullability contracts set by the delegate itself as lambda expressions return non-nulalble values as required by the delegate attributes.
Updating the compiler to the latest version shouldn't add new warnings since there are no actual problems in the code to begin with.

**Actual Behavior**:
Nullability mismatch warnings are reported for both lambda expression in the program above:
```
    warning CS8622: Nullability of reference types in type of parameter 's' of 'lambda expression' doesn't match the target delegate 'D1<T>' (possibly because of nullability attributes).
    warning CS8621: Nullability of reference types in return type of 'lambda expression' doesn't match the target delegate 'D2<T>' (possibly because of nullability attributes).
```

**Notes**:
It looks like Roslyn didn't verify nullability declared by delegate attributes at all and now it does, so sometimes the new warnings _might_ be correct where there was no warning at all previously.

Therefore, I'm not sure whether it's a bug or intended behavior but please consider the following:
- the attributes required to fix this warning without suppression simply aren't allowed in C# 9
- code written in C# 9 will have to deal with these new warnings even if it doesn't have any nullability problems upon updating a compiler version because the correct annotation simply wasn't possible in C# 9. E.g. an example above doesn't have any actual nullability issues
- in C# 9 you still cannot add the nullable attributes required by the warnings so if you cannot migrate to the latest C# as well you are required to either rewrite your code in a different way (e.g. replace lambda expression with method groups) or suppress the warnings

While it's useful to verify such contracts, perhaps it could be done by using target delegates' attributes instead of lambda expression attributes if there are none on the lambda expression? It would make code without nullability problems compile without warnings; it will not miss any nullability problems even if you don't annotate the lambda expression and even with C# 10 it will save some code duplication where you currently _have_ to duplicate every attribute on the lambda expression to make it compatible with the target delegate which feels redundant.

**Comment by jcouv** [comment]
We'd had a conversation along these lines with Rikki at the time of https://github.com/dotnet/roslyn/pull/56792. I'm not yet convinced, but we can discuss.
Now that attributes are allowed in lambdas, their presence/absence should be honored. The language doesn't have the concept of target-attributing. It's unfortunate that it can lead to redundancy in some scenarios, but the behavior falls out.
